Midships 4.7.0

yuxiang-midships opened the pull request:
Hi Ping SDK Team,

Yuxiang here from Midships. I am raising this pull request to show some changes we are doing to support some of our clients for Ping SDK integration. These change on iOS SDK is to allow certain feature available on the Android SDK to be also available on the iOS SDK.

Summary:
1. expose the token manager interface and instance on iOS sdk
2. expose the clear method in token manager on the iOS sdk
3. add the persist token method in token manager on the iOS sdk 

Feel free to reach out to me for more details and why some clients will need these changes.

Best regards
Yuxiang
Associate Director, Midships

spetrov committed:
Merge pull request #278 from ForgeRock/develop

ForgeRock iOS SDK 4.4.1 Release

spetrov committed:
Merge pull request #280 from ForgeRock/develop

Fixed privacy manifest file in the SDK's modules

spetrov committed:
Merge pull request #289 from ForgeRock/develop

ForgeRock iOS SDK 4.5.0 Release

vahancouver committed:
Merge pull request #310 from ForgeRock/develop

ForgeRock iOS SDK 4.6.0 Release

yuxiang-midships committed:
Expose TokenManager Instance via FRAuth
Similar to android SDK 's Config.getInstance().getTokenManager() to get token manager instance

yuxiang-midships committed:
Expose clearCredential method in Token Manager
Similar to Android SDK's TokenManger.clear() method

yuxiang-midships committed:
Add persistAccess token method in TokenManager
Similar to Android SDK's TokenManger.persist(AccessToken) method

george-bafaloukas-forgerock commented:
Hi @yuxiang-midships thanks for raising this PR. 

Can you please provide more information on the use cases you want to cover by making the Token manager and the "persist token method" public?

Commit:
Expose oAuth2 Client via token manager

yuxiang-midships commented:
https://github.com/ForgeRock/forgerock-ios-sdk/pull/320#issuecomment-2651054630

Hi George,

Thank you for taking a look at this. We are using FRSession instead of FRuser to have more control over different authentication tree through out the mobile apple CIAM journeys. At the same time, we have Oauth2 flows that will need to be supported. For the android SDK, we can get around and achieve the logic we want as we can access:

Config.getInstance().getSessionManager().getTokenManager()

which allow us to access on android the tokenManager.

We can then use the exchangeToken, clear and persist method in the token manager class to achieve using FR session for invoking AM trees and then use the tokenID to do oauth2 flows.

However when it come to the iOS SDK, we are not able to access the token manager and use the equivalent method. Therefore we modified the iOS SDK as illustrated in the commits to allow iOS sdk to have the same feature as the android SDK

We make the token manager and the ouath2 client in the token manager public. and we also make the clearCredential method in token manager public. Lastly we added the persist method to the token manager.

george-bafaloukas-forgerock commented:
Hi there,

Thanks for your reply. I will be actively reviewing this on the current Sprint. I am assuming this is related with the support case: `01076030` as well. Please expect more questions as part of that :)

george-bafaloukas-forgerock commented:
Hey Yuxiang and team, one more clarification. 
You mention
> We are using FRSession instead of FRuser to have more control over different authentication tree through out the mobile apple CIAM journeys

This is correct through the use of FRSession.authenticate you can control the use of the `authentication tree`. At the end of the authentication you can then jump on the FRUser classes. For example check out the following code:

`func handleNode(token: Token?, node: Node?, error: Error?) {
        if (token == nil) && (node == nil) && (error == nil) {
            print("All nil indicates a noSession=true completed Journey")
            self.updateStatus()
            return
        }
        if error != nil {
            print(error!)
            return
        }
        self.currentNode = node
        if let _ = token {
            print("Session only Journey is now complete")
            FRUser.currentUser?.getAccessToken(completion: { user, error in
                DispatchQueue.main.async {
                    self.updateStatus()
                }
            })
        } else {
            if node == nil {
                return
            }
            for callback: Callback in node!.callbacks {
            }
        }
    }`
So basically after getting the `Token` you can work with the FRUser interface which should expose all the methods required. Please let me know if I am missing something here. The only possible reason I can think of exposing more from the TokenManager would be the use of different OAuth2.0 flows other than the one the SDK supports OOTB. 

Same logic applies to Android

yuxiang-midships commented:
Hi George,

Thanks for reaching out. You are absolutely correct that after using the FRSession to finish the tree and get a SSO token, when can then use FR user to exchange the SSO token for Ouath2 Access token. However there are 2 issues we faced.

yuxiang-midships commented:
Hi George,

You are absolutely correct that after using FRSession, we can use FRUser to exchange for Oauth2 Access token. However two of our clients are also using AM trees to carryout extra authentication at a transaction level (not session level). In the flow we implemented, we have custom Ouath2 scripts that enable issuance of special scoped access token where the mobile end can use to invoke certain high risk transactions.

george-bafaloukas-forgerock commented on FRAuth/FRAuth/Manager/TokenManager.swift:
```diff
@@ -12,17 +12,17 @@ import Foundation
+public struct TokenManager {
     
-    var oAuth2Client: OAuth2Client
+    public var oAuth2Client: OAuth2Client
```
Add code doc for the public oAuth2Client

george-bafaloukas-forgerock commented on FRAuth/FRAuth/Manager/TokenManager.swift:
```diff
@@ -366,6 +366,10 @@ struct TokenManager {
         Browser.currentBrowser = nil
     }
     
+    public func persistToken(token: AccessToken) {
```
add code doc for the persistToken method

george-bafaloukas-forgerock reviewed (changes requested):
Hello @yuxiang-midships  and team, thank you very much for submitting this request and providing these clarifications. 

Can I please ask you to have a look on the comments and also make sure that any follow up commits are done using "Verified" account status. Please let us know if you need details on how to set up a "Verified" key.

Last but not least, can you please add some tests for the proposed changes?

Kind regards,
George Bafaloukas

vahancouver commented on FRAuth/FRAuth/Manager/TokenManager.swift:
```diff
@@ -12,17 +12,17 @@ import Foundation
 
 /// TokenManager class is a management class responsible for persisting, retrieving, and refreshing OAuth2 token(s)
-struct TokenManager {
+public struct TokenManager {
```
Please update the copyright header to reflect the latest changes year

vahancouver commented on FRAuth/FRAuth/FRAuth.swift:
```diff
@@ -50,7 +50,7 @@ public final class FRAuth: NSObject {
     var oAuth2Client: OAuth2Client?
     /// TokenManager instance for FRAuth to perform any token related operation
-    var tokenManager: TokenManager?
+    public var tokenManager: TokenManager?
```
Please update the copyright header to reflect the latest changes year

vahancouver commented on FRAuth/FRAuth/Manager/TokenManager.swift:
```diff
@@ -12,17 +12,17 @@ import Foundation
     ///   - oAuth2Client: OAuth2Client instance for OAuth2 token protocols
     ///   - keychainManager: KeychainManager instance for secure credentials management
-    public init(oAuth2Client: OAuth2Client, keychainManager: KeychainManager) {
+    init(oAuth2Client: OAuth2Client, keychainManager: KeychainManager) {
```
was public meant to be removed here?

vahancouver reviewed (changes requested):
Left some comments to address

PS: the commits need to be signed
